src/OrchardCore.Modules/OrchardCore.Media/Processing/MediaTokenService.cs

```diff
@@ -61,54 +60,75 @@ public string TokenizePath(string path)
+            if (otherCommands != null)
             {
-                entry.SlidingExpiration = TimeSpan.FromHours(5);
+                foreach (var command in otherCommands)
+                {
+                    processingCommands.Add(command.Key, command.Value);
+                }
+            }
 
-                var json = JsonConvert.SerializeObject(processingCommands);
+            return QueryHelpers.AddQueryString(pathParts[0], processingCommands);
+        }
+
+
+        public bool TryValidateToken(IDictionary<string, string> commands, string token)
+        {
+            GetHashKey();
 
-                return _dataProtector.Protect(json);
-            });
+            var commandValues = String.Concat(commands.Values);
 
-            otherCommands["token"] = queryStringToken;
+            var queryStringTokenKey = TokenCacheKeyPrefix + commandValues;
 
-            var commandCacheKey = CommandCacheKeyPrefix + key;
+            // Store a hash of the valid query string commands.
+            var queryStringToken = GetHash(commandValues, queryStringTokenKey);
 
-            var query = _memoryCache.GetOrCreate(commandCacheKey, entry =>
+            if (String.Equals(queryStringToken, token, StringComparison.OrdinalIgnoreCase))
             {
-                entry.SlidingExpiration = TimeSpan.FromHours(24);
-                return processingCommands;
-            });
+                return true;
+            }
+
+            return false;
 
-            return QueryHelpers.AddQueryString(pathParts[0], otherCommands);
         }
 
-        public IDictionary<string, string> GetTokenizedCommands(string token)
-        {
-            return _memoryCache.GetOrCreate(token, entry =>
-            {
-                entry.SlidingExpiration = TimeSpan.FromHours(24);
-                try
-                {
-                    var json = _dataProtector.Unprotect(token);
+        private string GetHash(string commandValues, string queryStringTokenKey)
+            => _memoryCache.GetOrCreate(queryStringTokenKey, entry =>
+               {
+                   entry.SlidingExpiration = TimeSpan.FromHours(5);
 
-                    return JsonConvert.DeserializeObject<IDictionary<string, string>>(json);
-                }
-                catch
-                {
-                }
+                   using var hmac = new HMACSHA256(_hashKey);
+
+                   var bytes = Encoding.UTF8.GetBytes(commandValues);
+
+                   var hash = hmac.ComputeHash(bytes);
+
+                   return Convert.ToBase64String(hash);
+               });
 
-                return null;
-            });
+        private void GetHashKey()
+        {
+            if (_hashKey == null)
+            {
+                var siteSettings = _siteService.GetSiteSettingsAsync().GetAwaiter().GetResult();
+                var mediaTokenSettings = siteSettings.As<MediaTokenSettings>();
+                _hashKey = mediaTokenSettings.HashKey;
```

Comment:
So @jtkech I moved it outside the constructor.

It's not totally on demand though, still held locally for the instance. 
I prefer this over making the methods async, and fully on demand, as it would break the `Orchard.AssetUrl` helper.

I left the migration for the moment, but it was intended as a one hit migration for existing sites. And again trying to avoid to much `.GetAwaiter()` work inside this service.

Could change that if you feel strongly about it ;)